proxy/hdrs/HTTP.cc

```diff
@@ -699,7 +699,19 @@ http_hdr_url_set(HdrHeap *heap, HTTPHdrImpl *hh, URLImpl *url)
     if (hh->u.req.m_url_impl != nullptr) {
       heap->deallocate_obj(hh->u.req.m_url_impl);
     }
-    hh->u.req.m_url_impl = url;
+    // Clone into new heap if the URL was allocated against a different heap
+    if ((char *)url < heap->m_data_start || (char *)url >= heap->m_free_start) {
```

Comment:
I think you might want to look at the implementation of `TSUrlClone` and follow that.

Reply:
That would be nice the body of the if here would be replaced by url_copy(), but that call requires the source heap, and I don't know how to get that from the UrlImpl object.